File: Game.Application/GameJS/src/terrain-generation/TerrainPoint.ts
import { TemperatureCategory } from "./TemperatureCategory";
import { HumidityCategory } from "./HumidityCategory";
import { BiomeDetails } from "./BiomeDetails";
import { AltitudeCategory } from "./WaterCategory";
import { TerrainSettings } from "./TerrainSettings";
import { VisualTerrainType } from "./VisualTerrainType";

function indexOrLength<T extends number>(steps: number[], value: number): T {
  const result = steps.findIndex(
    v => v > value
  );
  return (result === -1 ? steps.length : result) as T;
}

export class TerrainPoint {
  public readonly terrainSettings: TerrainSettings;
  public readonly x: number;
  public readonly y: number;
  public readonly altitude: number;
  public readonly heat: number;
  public readonly humidity: number;
  public readonly feature: number;
  public readonly caveIndicator: number;
  constructor(
    terrainSettings: TerrainSettings,
    x: number,
    y: number,
    altitude: number,
    heat: number,
    humidity: number,
    feature: number,
    caveIndicator: number
  ) {
    this.terrainSettings = terrainSettings;
    this.x = x;
    this.y = y;
    this.altitude = altitude;
    this.heat = heat;
    this.humidity = humidity;
    this.feature = feature;
    this.caveIndicator = caveIndicator;
  }
  get temperatureCategory() {
    return indexOrLength<TemperatureCategory>(this.terrainSettings.tempsStep, this.heat);
  }
  get humidityCategory() {
    return Math.min(
      BiomeDetails.biomeLabels[this.temperatureCategory].length,
      indexOrLength<HumidityCategory>(this.terrainSettings.humidityStep, this.humidity)
    );
  }
  get biomeLabel() {
    return BiomeDetails.biomeLabels[this.temperatureCategory][
      this.humidityCategory
    ];
  }
  get biomeCategory() {
    return BiomeDetails.categoryLookup[this.biomeLabel];
  }
  get altitudeCategory() {
    return indexOrLength<AltitudeCategory>(this.terrainSettings.altitudeStep, this.altitude);
  }
  get visualCategory(): VisualTerrainType {
    return (this.terrainSettings.visualizationSpec).execute({
      altitude: this.altitude,
      heat: this.heat,
      humidity: this.humidity,
      feature: this.feature,
      temperatureCategory: this.temperatureCategory,
      humidityCategory: this.humidityCategory,
      biomeCategory: this.biomeCategory,
      altitudeCategory: this.altitudeCategory,
    });
  }
  get detailVisualCategory(): VisualTerrainType {
    return (this.terrainSettings.detailVisualizationSpec).execute({
      altitude: this.altitude,
      heat: this.heat,
      humidity: this.humidity,
      feature: this.feature,
      temperatureCategory: this.temperatureCategory,
      humidityCategory: this.humidityCategory,
      biomeCategory: this.biomeCategory,
      altitudeCategory: this.altitudeCategory,
    });
  }
  get hasCave() {
    return (
      (this.altitudeCategory === AltitudeCategory.Hills ||
        this.altitudeCategory === AltitudeCategory.None) &&
      this.caveIndicator >= 3e6 && this.caveIndicator <= 6e6
    );
  }
}

File: Game.Domain/Terrain/TerrainPoint.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;

namespace Game.Domain.Terrain
{
    public readonly struct TerrainPoint : ITerrainSituation
    {
        public readonly GameCoordinate coordinates;
        public readonly double humidity;
        public readonly double heat;
        public readonly double altitude;
        public readonly double feature;
        private readonly bool isCave;
        public readonly TemperatureCategory temperatureCategory;
        public readonly HumidityCategory humidityCategory;
        public readonly AltitudeCategory altitudeCategory;

        public TerrainPoint(GameCoordinate coordinates, double humidity, double heat, double altitude, double feature, bool isCave, TerrainSettings settings)
        {
            this.coordinates = coordinates;
            this.humidity = humidity;
            this.heat = heat;
            this.altitude = altitude;
            this.feature = feature;
            this.isCave = isCave;
            temperatureCategory = TerrainSettings.GetBestKey(settings.TemperatureStep, heat);
            humidityCategory = (HumidityCategory)Math.Min(
                  BiomeDetails.biomeLabels[temperatureCategory].Count - 1, 
                  (int)TerrainSettings.GetBestKey(settings.HumidityStep, humidity)
                );
            altitudeCategory = TerrainSettings.GetBestKey(settings.AltitudeStep, altitude);

            //System.Diagnostics.Debug.Assert(temperatureCategory >= TemperatureCategory.Polar && temperatureCategory <= TemperatureCategory.Tropical);
            //System.Diagnostics.Debug.Assert(humidityCategory >= HumidityCategory.Superarid && humidityCategory <= HumidityCategory.Superhumid);
            //System.Diagnostics.Debug.Assert(altitudeCategory >= AltitudeCategory.DeepWater && altitudeCategory <= AltitudeCategory.Mountain);

        }

        public BiomeLabel BiomeLabel => BiomeDetails.biomeLabels[temperatureCategory][humidityCategory];
        public BiomeCategory BiomeCategory => BiomeDetails.CategoryLookup[BiomeLabel];

        public double Altitude => altitude;

        public double Heat => heat;

        public double Humidity => humidity;

        public double Feature => feature;

        public TemperatureCategory TemperatureCategory => temperatureCategory;

        public HumidityCategory HumidityCategory => humidityCategory;

        public AltitudeCategory AltitudeCategory => altitudeCategory;

        public bool IsCave => isCave;
    }
}